xocto/ranges.py

```diff
@@ -221,6 +236,14 @@ def __str__(self) -> str:
     def __repr__(self) -> str:
         return f"<Range: {str(self)}>"
 
+    def __setattr__(self, name: str, value: Any) -> None:
+        if name in type(self).__slots__:
+            raise AttributeError("Can't set attributes")
+        else:
+            raise AttributeError(
+                f"'{type(self).__name__}' object has no attribute '{name}'"
+            )
```

Comment:
Perhaps explicit is better than implicit, but I wonder if we could delegate this branch to `super()`?